### ISSUE TITLE
Training pipeline

### ISSUE THREAD

**Issue by lukeforehand** [lukeforehand/deertracker#5]
Hoping to add a cli entrypoint here:
https://github.com/lukeforehand/deertracker/blob/main/deertracker/main.py

I'm not familiar with setting up a training pipeline, but before training it will be assumed that photos have already been labeled and separated into a training and test set.

**Comment by kbrose** [comment]
I'll work on getting an MR up for this in the next few days. It's common for classification tasks to organize the data into folders by label:

```
train/
  doe/
    img1.jpg
    ...
  buck/
  spike/
  squirrel/
  ...
test/
  doe/
  ...
```

so I'll assume a structure like that for the first pass implementation.

Do you have a machine with an nvidia GPU? It's not required for training but usually gives between a 10-40x speedup. Alternatively, apple just released a [version of tensorflow that works with their Metal API](https://github.com/apple/tensorflow_macos) which I haven't tried yet but might be an option for you.

**Comment by lukeforehand** [comment]
That’s great!  I have an older Nvidia GPU but I’ll have to revive the machine :-). I was wondering if these training images should be cropped to the object or should they be the original image with bounding box info in another file?

Another question is, does training enhance the source model’s detection capabilities or are we just teaching it about new classes?

**Comment by lukeforehand** [comment]
Another question:  can we create more training data by distorting images and by extracting frames from video?  I have way more photos of deer than say owls but I would like to try classifying all the wildlife we see.

Also I read that google colab lets you use a free GPU in your notebook, and files can be retrieved from your google drive.  Do you think that is worth looking into?  I have yet to see how much disk space these photos use.  Should I resize them in a certain way so our training dataset can be archived and shared around more easily?

**Comment by kbrose** [comment]
> I was wondering if these training images should be cropped to the object or should they be the original image with bounding box info in another file?

I'm not sure it makes sense to go through the effort of labeling bounding boxes. It takes more time to label, and it also makes the model more complicated (both in runtime and code complexity). Typically bounding boxes are useful when the image has lots of objects in it, do a lot of the images show multiple animals?

> Another question is, does training enhance the source model’s detection capabilities or are we just teaching it about new classes?

There's some nuance here, but it'd be mostly accurate to think of it more as just teaching the pretrained model about new classes.

There's a couple different ways to improve on a pre-trained model. The simplest approach is to keep most of the pre-trained model static, and replace the last few layers with your own which will then get trained. So the part of your model which is the "source model" isn't really improving, but the model as a whole will work better (or work at all, if you need to teach it new classes) on the task at hand.

Another approach is to not hold the model static, so that you are simultaneously training a new set of layers and shifting the weights of the source model's layers. This usually takes more time to get right, though.

> Another question: can we create more training data by distorting images and by extracting frames from video? I have way more photos of deer than say owls but I would like to try classifying all the wildlife we see.

Yes, this is very standard. Typically it's done "online", i.e. we don't pre-generate these distorted images but do it on the fly during the training process. See e.g. [`tf.image.random_brightness`](https://www.tensorflow.org/api_docs/python/tf/image/random_brightness) and the other `tf.image.random_*` operators that include flipping, rotating, shifting the hue, etc..

> Also I read that google colab lets you use a free GPU in your notebook, and files can be retrieved from your google drive. Do you think that is worth looking into?

I've heard good things about colab, but I've never used it personally. Probably worth playing around.

> I have yet to see how much disk space these photos use. Should I resize them in a certain way so our training dataset can be archived and shared around more easily?

I'd wait to do any modification until we see how much space they take up. It's pretty standard to use a much smaller image size than we're used to viewing, typically in the range of 256x256 or 224x224. But again, no need to downsize until we know it's necessary. Having extra pixels around can make the augmentation described above more robust (i.e. you have more room to crop a random square from the image, etc.)

**Comment by lukeforehand** [comment]
> I'm not sure it makes sense to go through the effort of labeling bounding boxes. It takes more time to label, and it also makes the model more complicated (both in runtime and code complexity). Typically bounding boxes are useful when the image has lots of objects in it, do a lot of the images show multiple animals?

Yes and maybe this will be the challenging part of the project; the deer are often grouped together in photos and deer and turkeys as well.  One of my hopes was the ability to count number of turkeys or does in a time frame to be able to track a group of animals over time.  A common situation is I see three different groups of turkeys in a morning, a 20+ turkey group followed by a 10+ turkey group followed by a 5 group of huge Tom turkeys.  Perhaps we can get more general in the classification (like "turkey group") or ("deer herd") when there are big groups?  My hope was achieving the ability to track animal classes and numbers over time and location.

Thank you for answering all of the other questions, it really clears a lot of unknowns for me :-)

**Comment by lukeforehand** [comment]
Can you read this and let me know what you think?  

https://github.com/microsoft/CameraTraps/blob/master/megadetector.md

It seems like a promising detector that we could start with, because it was trained specifically on trap photos to begin with and the code is easy to run with.

**Comment by lukeforehand** [comment]
In reading about mega detector I see it used this public dataset that is already annotated and contains a wider variety of species than just deer and turkeys :-). I’d like to catch owls, coyotes, wolves, wildcats so this could be promising additional dataset to my photos

https://beerys.github.io/CaltechCameraTraps/

![Image](https://i.imgur.com/1NouXhH.jpg)

**Comment by kbrose** [comment]
Yeah that could be a great resource to use. Looks like lots of examples for most classes which is important.

I put up an MR for the mega detector: #7 

![](https://user-images.githubusercontent.com/4053170/100376643-815fa280-2fd5-11eb-9c81-b2cc9ff9617d.png)
)

**Comment by lukeforehand** [comment]
Great!  I added some example outputs from your code in the README.md.  It's pretty cool to me that it detected an animal in the brush alongside another animal in the foreground!

**Comment by lukeforehand** [comment]
I’ve been running some tests and I am very impressed with the detection!  It seems very accurate in detecting the animal class.  Given this, it seems possible to run images through the detector to create many animal crops.  Do you think this could be a very easy way to generate training images, because then a labeled only has to identify the class of the animal crop.  Then would that be sufficient training data for tweaking the last layers in order to add all the classes?  Also we can incorporate the CalTech data set in the same way but instead using the annotations to extract the crops.  Tell me if this is a bad idea :-). I imagine the entire photo provides context for the model but unsure if that is mainly for the detector and not the classifier.

**Comment by kbrose** [comment]
Yeah that's exactly what I was thinking, including using the annotations to extract crops. I think they should be sufficient contact for the animal classification task.

My suggestion would be to extract a slightly larger crop than what the model/annotation says. Maybe something like 1% of the image's width/height or 10 pixels (whichever is larger) on all sides. Then during training we can take out a random subcrop to make it robust against a misbehaving megadetector during inference.

**Comment by lukeforehand** [comment]
I’m excited!  My photo import code guards against processing the same photo twice so I can modify it to produce the bigger crop as you suggest.  I’ll start processing my google drive dataset soon!  Do you want access to the crop results as well?  

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

**Comment by lukeforehand** [comment]
If the import process does or has problems I can resume it so that’s nice.

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

**Comment by lukeforehand** [comment]
*dies

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

**Comment by lukeforehand** [comment]
@kbrose would you be willing to look at these 34 images ?  There are two otter-like animals I'm unsure of, and I can't tell the difference between a fox and a coyote.
https://drive.google.com/file/d/1WKeSq34bzFRONX9qll9sGaQoxrxwYOyN/view?usp=sharing

**Comment by lukeforehand** [comment]
I have labeled my "whitetail" dataset and am now going through the bucks to further differentiate between spike/nubbin and buck.

**Comment by lukeforehand** [comment]
> @kbrose would you be willing to look at these 34 images ? There are two otter-like animals I'm unsure of, and I can't tell the difference between a fox and a coyote.
> https://drive.google.com/file/d/1WKeSq34bzFRONX9qll9sGaQoxrxwYOyN/view?usp=sharing

Nevermind, got some expert advice from Dad :-)

Okay the labeled crops "whitetail" dataset is complete:


[Download crops](https://drive.google.com/file/d/19qN5PnY8PeqQ2mQ6PFyi41wuu7NgLKhH/view?usp=sharing)

3285 total crops
Distribution:

```bash
1402 buck
 822 bear
 330 doe
 232 turkey
 166 person
 113 yearling_buck
 110 fawn
  21 vehicle
  21 coyote
  20 porcupine
  15 crow
  12 fisher
  12 dog
   6 sandhill_crane
   6 mountain_lion
   4 raccoon
   4 blue_jay
   3 skunk
   3 bobcat
   3 bird
```

**Comment by lukeforehand** [comment]
@kbrose Training on google colab, need to write results out but a good sign, it is blowing through epochs compared to my laptop!

https://colab.research.google.com/drive/1P1OowND63_mFAe0kpoXyFIIzAkcHTS-G?usp=sharing

**Comment by lukeforehand** [comment]
on the whitetail crops:
```
Epoch  495 (0.50s), Loss: 0.00, Accuracy: 100.00, Test Loss: 1.27, Test Accuracy: 85.57
Epoch  496 (0.51s), Loss: 0.00, Accuracy: 100.00, Test Loss: 1.34, Test Accuracy: 84.54
Epoch  497 (0.50s), Loss: 0.00, Accuracy: 100.00, Test Loss: 1.37, Test Accuracy: 85.57
Epoch  498 (0.51s), Loss: 0.00, Accuracy: 100.00, Test Loss: 1.37, Test Accuracy: 85.57
Epoch  499 (0.51s), Loss: 0.00, Accuracy: 100.00, Test Loss: 1.37, Test Accuracy: 85.57
Epoch  500 (0.50s), Loss: 0.00, Accuracy: 100.00, Test Loss: 1.38, Test Accuracy: 85.57
```

**Comment by lukeforehand** [comment]
I fixed the crop tar it had hidden files in it, the notebook works if you wanted to give colab a go @kbrose.  I was also thinking I could work on filtering out classes without a minimum # images if you wanted me to.

**Comment by kbrose** [comment]
I've got the filtering working in my latest [PR](https://github.com/lukeforehand/deertracker/pull/15), should be ready to merge tonight, just a couple points left I want to finish on it.
